Api tests

AlexandraNayanzina opened the pull request:
- BuildTypeTestHomeWork.java - 2 tests from the course + 2 tests on roles
- ProjectTest.java - tests on creating project

AlexandraNayanzina committed:
Add config, spec, base tests

AlexandraNayanzina committed:
Add crud interface, endpoint, request,

AlexandraNayanzina committed:
Test "User should be able to create a build".
Added Steps:
Create user, Create project, Create build type, created build type assertions

AlexandraNayanzina committed:
Added TestData class, added negative test,

AlexandraNayanzina committed:
Added TestDataStorage. Implemented removing created data

AlexandraNayanzina committed:
Added BuildTypeTestHomeWork class with 4 tests.
2 tests - buildType creation (that were explained on the Module)
2 tests - for the projectAdmin role (Home Work)

AlexandraNayanzina committed:
Added ProjectTest class with tests on project creation

AlexandraNayanzina committed:
Delete request-examples/buildConfigurationTest.http

nobugs-me commented on src/test/java/org/example/teamcity/api/BuildTypeTestHomeWork.java:
```diff
@@ -0,0 +1,136 @@
+        response.then()
+                .assertThat()
+                .statusCode(SC_BAD_REQUEST)
+                .body(containsString("The build configuration / template ID \"%s\" is already used by another configuration or template".formatted(testData.getBuildType().getId())));
```
В будущем все одинаковые типы ассертов можно вынести в отдельный класс и добавлять набор ассертов как ResponseSpecification

AlexandraNayanzina replied:
Fixed

nobugs-me commented on src/test/java/org/example/teamcity/api/BuildTypeTestHomeWork.java:
```diff
@@ -0,0 +1,136 @@
+
+        // SuperUser creates an ProjectAdmin user assigned to the project
+        var projectAdmin = testData.getUser();
+        var roles = generate(Roles.class, "PROJECT_ADMIN", "p:" + testData.getProject().getId());
```
"PROJECT_ADMIN" можно вынести в Enum

AlexandraNayanzina replied:
Fixed

nobugs-me commented on src/test/java/org/example/teamcity/api/BuildTypeTestHomeWork.java:
```diff
@@ -0,0 +1,136 @@
+        requester.getRequest(BUILD_TYPES).create(buildType1);
+
+        // WHEN: the user creates second buildType
+        var testData2 = generate();
```
Обычно используют нейминг не в числах, а в словах - firstTestData и тд

AlexandraNayanzina replied:
Thank you!

AlexandraNayanzina replied:
Fixed

nobugs-me reviewed (approved):
Привет! 

Спасибо за выполнение базового ДЗ! ⚡️

Тест выполнен корректно, поэтому ставлю Approve 🤝

AlexandraNayanzina committed:
The comments from Alexa were fixed:
1. Enum is created for the Roles
2. ResponseSpecification is added
3. naming is fixed (e.g. testData2-> secondTestData)

AlexandraNayanzina committed:
Polishing
